BigAnimal - AWS Secrets Manager Integration (#5465)
EnterpriseDB/docs: product_docs/docs/biganimal/release/administering_cluster/aws_secrets_manager_integration.mdx

   
   client = boto3.client('secretsmanager', region_name=region)
   response = client.get_secret_value(
               SecretId='dev/toy/demo'
                   )
   
   secretDict = json.loads(response['SecretString'])
   
   connection = psycopg2.connect(
            user=secretDict['username'], 
             password=secretDict['password'],
              host=secretDict['host'], 
                port=secretDict['port'],
                sslmode='require',
                 database=secretDict['dbname'])
   
   mycursor = connection.cursor()
   
   create = "create table Demo0503(Toyota int)"
   #sql = "INSERT into secretmgr(id,name) values (%s, %s)"
   #value = (2, "Toyota_Demo")
   mycursor.execute(create)
   
   connection.commit()
   ```

  Fetch all the rows from the database:

   ```shell
   print(mycursor.rowcount, "record")
   ```

## Example

In this example, a script file has all the commands required to create a Secrets Manager rotation Lambda function, execute the rotation script, and execute the sample application.

```shell
cat gen_pass_rotate_bigani_and_secretmgr_pass.py 
import secrets
import string
import requests
import json
import boto3

> key = "baak_1tbgf3PXdx2L7nclsMHAdGqHy9bNNAZtqgj3Y2DeYltjXkP5n"

Author: ```suggestion
key = "<your_access_key>"
```

---
title: "AWS Secrets Manager integration"
---

With BigAnimal, you can use AWS Secrets Manager. AWS Secrets Manager helps you manage, retrieve, and rotate database credentials, access keys, and other secrets throughout their lifecycle.

To create a secret manager:

1. Create a PostgreSQL cluster on the BigAnimal portal.

1. Create and save an [access key](../reference/access_key/#create-your-personal-access-key).

1. Create a secret in AWS Secrets Manager for your psql credentials. 

   Create the secret manager using a Lambda script or using the AWS console:

   - Lambda script:

     ```shell
     import boto3
     import json

     def create_secret(secret_name, username, password, database, host):
                client = boto3.client('secretsmanager')

                secret_string = json.dumps({
           "username": username,
           "password": password,

>            "engine": "mysql",

Author: ```suggestion
           "engine": "postgresql",
```

           "host": host,
           "dbname": database,

>            "port": 3306

Author: ```suggestion
           "port": 5432
```

1. Create the secret in the centralized Secrets Manager for your access key.

1. Create a sample login application.

   For example, using a Lambda script:

   ```shell
   [cloudshell-user@ip-10-130-83-78 ~]$ cat lambda_connect.py 
   import json
   import boto3
   import base64
   import psycopg2
   region = 'us-east-1'
   
   client = boto3.client('secretsmanager', region_name=region)
   response = client.get_secret_value(
               SecretId='dev/toy/demo'
                   )
   
   secretDict = json.loads(response['SecretString'])
   
   connection = psycopg2.connect(
            user=secretDict['username'], 
             password=secretDict['password'],
              host=secretDict['host'], 
                port=secretDict['port'],
                sslmode='require',
                 database=secretDict['dbname'])
   
   mycursor = connection.cursor()
   
   create = "create table Demo0503(Toyota int)"
   #sql = "INSERT into secretmgr(id,name) values (%s, %s)"
   #value = (2, "Toyota_Demo")
   mycursor.execute(create)
   
   connection.commit()
   ```

  Fetch all the rows from the database:

   ```shell
   print(mycursor.rowcount, "record")
   ```

> 1. Create a Secrets Manager rotation Lambda function.

Author: ```suggestion
1. Create a Secrets Manager rotation Lambda function.
```

> 1. Manually execute the Secrets Manager rotation script.

Author: ```suggestion
```

> 1. Manually execute your sample application.

Author: ```suggestion
```

## Example

In this example, a script file has all the commands required to create a Secrets Manager rotation Lambda function, execute the rotation script, and execute the sample application.

```shell
cat gen_pass_rotate_bigani_and_secretmgr_pass.py 
import secrets
import string
import requests
import json
import boto3

> key = "baak_1tbgf3PXdx2L7nclsMHAdGqHy9bNNAZtqgj3Y2DeYltjXkP5n"

Author: Do not write this exact key.

---
title: "AWS Secrets Manager integration"
---

> With BigAnimal, you can use AWS Secrets Manager. AWS Secrets Manager helps you manage, retrieve, and rotate database credentials, Access keys, and other Secrets throughout their lifecycle.

Reviewer: ```suggestion
With BigAnimal, you can use AWS Secrets Manager. AWS Secrets Manager helps you manage, retrieve, and rotate database credentials, access keys, and other secrets throughout their lifecycle.
```

To create a secret manager:

> 1. Create a PostgreSQL cluster on BigAnimal portal.

Reviewer: ```suggestion
1. Create a PostgreSQL cluster on the BigAnimal portal.
```

> 1. Create and save an [Access key](../reference/access_key/#create-your-personal-access-key).

Reviewer: ```suggestion
1. Create and save an [access key](../reference/access_key/#create-your-personal-access-key).
```

> 1. Create secret in application Secrets Manager for your **psql** credentials. 

Reviewer: ```suggestion
1. Create a secret in AWS Secrets Manager for your psql credentials. 
```
I think this is what was meant?

   Create the secret manager using a Lambda script or using the AWS console:

   - Lambda script:

     ```shell
     import boto3
     import json

     def create_secret(secret_name, username, password, database, host):
                client = boto3.client('secretsmanager')

                secret_string = json.dumps({
           "username": username,
           "password": password,
           "engine": "mysql",
           "host": host,
           "dbname": database,
           "port": 3306
         })

    response = client.create_secret(
        Name=secret_name,
        SecretString=secret_string
    )

    return response
    ```
    

>     Using the created secret:

Reviewer: Does this mean: To use the created secret?

    ```shell
    create_secret('mySecretName', 'myUsername', 'myPassword', 'myDatabase', 'myHost')
    ```

   - AWS console:
   

>      1. Search for Secret Manager under services.

Reviewer: Should services be upper case and bold? Is it a UI element?

Reviewer 2: yes it is a UI element and should be capitalized

     1. Select **Store a new secret**.

>      1. Select **Credentials for other databases** on **Choose secret type** page and provide:

Reviewer: ```suggestion
     1. On the **Choose secret type** page, select **Credentials for other databases**  and provide:
```

        - Username
        - Password
        - Encryption key
        - Database

>           Provide the server address, database name and port as per the selected database engine. Select **Next**.

Reviewer: Should this be a step? Or is this part of the Database bullet? If the latter, then Select Next should be on its own line or its own step. Couldn't make out what was intended here.

Reviewer 2: it's part of the database bullet, so Select Next should be its own step

>      1. Provide **Secret name** on the **Configure secret** page. Optionally you can provide:

Reviewer: ```suggestion
     1. On the **Configure secret** page, provide **Secret name**. Optionally, you can provide:
```

        - Description
        - Tags
        - Resource permissions
        - Replicate secret
        Select **Next**.

>      1. Optionally, provide details on the **Configure rotation** page.

Reviewer: ```suggestion
     1. Optionally, on the **Configure rotation** page, provide details.
```

>      1. Review the code in different languages like: Java, JavaScript, C#, Python3, Ruby, Go and Rust. Select **Store** to create the secret manager.

Reviewer: ```suggestion
     1. Review the code in different languages like: Java, JavaScript, C#, Python3, Ruby, Go, and Rust. To create the secret manager, select **Store**.
```

> 1. Create secret in centralized Secrets Manager for your Access key.

Reviewer: ```suggestion
1. Create the secret in the centralized Secrets Manager for your access key.
```
Are these articles correct? Should it be "the" or "a" (in both cases)?

> 1. Create a sample login application:

Reviewer: ```suggestion
1. Create a sample login application.
```

> 1. Create a sample login application:

Reviewer: ```suggestion
1. Create a sample login application.
```

>    For example using Lambda script,

Reviewer: ```suggestion
   For example, using a Lambda script:
```

   ```shell
   [cloudshell-user@ip-10-130-83-78 ~]$ cat lambda_connect.py 
   import json
   import boto3
   import base64
   import psycopg2
   region = 'us-east-1'
   
   client = boto3.client('secretsmanager', region_name=region)
   response = client.get_secret_value(
               SecretId='dev/toy/demo'
                   )
   
   secretDict = json.loads(response['SecretString'])
   
   connection = psycopg2.connect(
            user=secretDict['username'], 
             password=secretDict['password'],
              host=secretDict['host'], 
                port=secretDict['port'],
                sslmode='require',
                 database=secretDict['dbname'])
   
   mycursor = connection.cursor()
   
   create = "create table Demo0503(Toyota int)"
   #sql = "INSERT into secretmgr(id,name) values (%s, %s)"
   #value = (2, "Toyota_Demo")
   mycursor.execute(create)
   
   connection.commit()
   ```

>   Fetch all the rows from database:

Reviewer: ```suggestion
  Fetch all the rows from the database:
```

   ```shell
   print(mycursor.rowcount, "record")
   ```

> 1. Create Secrets Manager rotation lambda function.

Reviewer: ```suggestion
1. Create a Secrets Manager rotation Lambda function.
```
I think Lambda should be capitalized as it is elsewhere? This is the name of a programming language?

> 1. Manually execute Secrets Manager rotation script.

Reviewer: ```suggestion
1. Manually execute the Secrets Manager rotation script.
```
Should this be "a" or "the"?

1. Manually execute your sample application.

## Example

> In this example a script file has all the commands required to create Secrets Manager rotation lambda function, execute the rotation script, and execute the sample application.

Reviewer: ```suggestion
In this example, a script file has all the commands required to create a Secrets Manager rotation Lambda function, execute the rotation script, and execute the sample application.
```
